### DISCUSSION TITLE
4.3.4 스프링 시큐리티 설정 테스팅 과정에서 403 오류가 납니다.

### DISCUSSION THREAD

**Discussion by seonkyungkim** [fsoftwareengineer/todo-application#49] [질문과 답변] [answered]
안녕하세요, 작가님.
프로그래밍 초보인데도 불구하고 참 재미있게 책을 읽어나가고 있습니다.
지금까지는 탈 없이 이해-구현-테스팅이 진행되었는데, 4.3.4 스프링 시큐리티 설정에서 문제가 생겼습니다.

테스팅 1, 2단계인 signup-signin-로그인 후 토큰확인까지 됐는데요.
로그인토큰을 복사해 GET 메서드로 요청하는 과정에서 결과물이 아무것도 뜨지 않습니다.

<img width="907" alt="Screen Shot 2022-01-05 at 8 34 36 PM" src="https://user-images.githubusercontent.com/57647921/148212144-dc85248d-2e44-4cf1-a5a0-3354207dfddd.png">

<img width="1255" alt="Screen Shot 2022-01-05 at 8 35 19 PM" src="https://user-images.githubusercontent.com/57647921/148212167-e5d302f0-7b80-4143-93ea-537734322c44.png">

인텔리제이에서는 애플리케이션도 에러없이 실행되고, 콘솔에도 오류로그는 찍히지 않았고 단순히 아래의 정보성 안내만 나타났습니다.
INFO 13654 --- [nio-8080-exec-9] c.e.t.security.JwtAuthenticationFilter   : Filter is running..

+ 추가
스프링부트 2.5.7 / 그래들 6.8.2 / jsonwebtoken 0.9.1 으로 맞춰둔 상태입니다.

지금은 아래처럼 403 에러가 명시적으로 보입니다. 토큰 뒤에 문자열을 덧붙여도 위조된 토큰 로그가 뜨지 않습니다. 뭐가 문제일까요?
<img width="921" alt="Screen Shot 2022-01-05 at 11 16 15 PM" src="https://user-images.githubusercontent.com/57647921/148232167-29ea49d6-e039-47fa-9b88-66d07ac8b087.png">

**Comment by fsoftwareengineer** [comment, marked as answer]
에러 메시지가 보이지 않는 경우 https://github.com/fsoftwareengineer/todo-application/discussions/41 에 관련된 내용이 아닌지 확인 해 보시면 좋을 것 같습니다.

위조된 토큰의 경우 403 forbidden이 리턴되는것이 정상입니다. 혹시 정상 토큰을 사용했는데도 403에러를 받으신건가요?

**Reply by seonkyungkim** [reply]
네. 정상토큰을 bearer token에 넣고 보냈는데도 403 Forbidden이 뜹니다. /auth/signin에 email과 password를 보내고 로그인 시도시에 결과로 돌아오는 Token 값을 복사해서 정상토큰으로 쓰고 있습니다. 혹시 제가 잘못된 토큰을 사용하고 있는건가요?

**Reply by fsoftwareengineer** [reply]
이미 패스워드 암호화를 하신 상태라면 https://github.com/fsoftwareengineer/todo-application/discussions/3 를 참고 해 보시기 바랍니다.

패스워드 암호화를 진행하지 않으신 상태라면 TokenProvider의 로직이 토큰을 올바르게 해석하지 못하는 것일 수 있습니다. 디버깅을 더 하기 위해서

1. 현재 가지고 있는 토큰을 base64로 디코딩 해 주세요. https://amp.base64decode.org
2. TokenProvider.java 와 WebSecurityConfig.java 전문을 (import 포함) 리뷰 해 주세요.
3. Filtering이 일어나는 부분에 라인마다 로그를 넣어 어디까지 실행되는지 확인 해 주세요.

1과 2를 이 쓰레드에 남겨주시면 저도 리뷰 해 보도록 하겠습니다.

**Reply by seonkyungkim** [reply]
jsonwebtoken 버전이 0.9.1임에도 불구하고 #48 을 참고해서 SECRET_KEY를 40자 이상으로 바꿔보았더니, JwtAuthenticationFilter의 과정마다 로그도 잘 찍히고 정상적으로 동작하는 듯 합니다.
<img width="458" alt="Screen Shot 2022-01-06 at 6 24 07 PM" src="https://user-images.githubusercontent.com/57647921/148360728-3a749301-9340-4925-9654-8058371b279d.png">

2번의 TokenProvider와 WebSecurityConfig는 에러발생시에 작가님의 코드를 그대로 가져와 경로만 바꾸어 설정해두었기 때문에 여기서 문제가 발생한 것 같지는 않습니다. 참고로 디코딩한 토큰은 아래와 같습니다.
<img width="745" alt="Screen Shot 2022-01-06 at 6 24 22 PM" src="https://user-images.githubusercontent.com/57647921/148360885-5b98506e-e67d-4937-af26-480ab71f88d4.png">


친절한 답변 감사합니다!
